src/theory/strings/output_channel.h

```diff
@@ -0,0 +1,308 @@
+   */
+  void doPendingFacts();
+  /** Do pending lemmas
+   *
+   * This method flushes all pending lemmas (d_pending_lem) to the output
+   * channel of theory of strings.
+   *
+   * Like doPendingFacts, this function will terminate early if a conflict
+   * has already been encountered by the theory of strings. The vector
+   * d_pending_lem is cleared regardless of whether a conflict is discovered.
+   *
+   * Notice that as a result of the above design, some lemmas may be "dropped"
+   * if a conflict is discovered in between when a lemma is added to the
+   * pending vector of this class (via a sendInference call). Lemmas
+   * e.g. those that are required for initialization should not be sent via
+   * this class, since they should never be dropped.
+   */
+  void doPendingLemmas();
+  /**
+   * Have we processed an inference during this call to check? In particular,
+   * this returns true if we have a pending fact or lemma, or have encountered
+   * a conflict.
+   */
+  inline bool hasProcessed() const
+  {
+    return hasConflict() || !d_pending_lem.empty() || !d_pending.empty();
+  }
+  /** Do we have a pending fact to add to the equality engine? */
+  inline bool hasPendingFact() const { return !d_pending.empty(); }
+  /** Do we have a pending lemma to send on the output channel? */
+  inline bool hasPendingLemma() const { return !d_pending_lem.empty(); }
+  /** Are we in conflict? */
+  bool hasConflict() const;
+
+ protected:
+  /**
+   * Indicates that ant => conc should be sent on the output channel of this
+   * class. This will either trigger an immediate call to the conflict
+   * method of the output channel of this class of conc is false, or adds the
+   * above lemma to the lemma cache d_pending_lem, which may be flushed
+   * later within the current call to TheoryStrings::check.
+   *
+   * The argument c is a string identifying the reason for inference, used for
+   * debugging.
+   */
+  void sendLemma(Node ant, Node conc, const char* c);
+  /**
+   * Indicates that conc should be added to the equality engine of this class
+   * with explanation eq_exp. It must be the case that eq_exp is a (conjunction
+   * of) literals that each are explainable, i.e. they already hold in the
+   * equality engine of this class.
+   */
+  void sendInfer(Node eq_exp, Node eq, const char* c);
+
+ private:
+  /** the parent theory of strings object */
+  TheoryStrings& d_parent;
+  /** the equality engine
+   *
+   * This is a reference to the equality engine of the theory of strings.
+   */
+  eq::EqualityEngine& d_ee;
+  /** the output channel
+   *
+   * This is a reference to the output channel of the theory of strings.
+   */
+  OutputChannel& d_out;
+  /** Common constants */
+  Node d_true;
+  Node d_false;
+  /** The list of pending literals to assert to the equality engine */
+  std::vector<Node> d_pending;
+  /** A map from the literals in the above vector to their explanation */
+  std::map<Node, Node> d_pending_exp;
```

Comment:
Since this is a new class, can we rename the variables to be camelCase to follow the style guide?